File: 187. Repeated DNA Sequences.cpp
#include"LeetCodeLibs.h"

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        unordered_set<string> res;
        unordered_set<string_view> seen;
        const string_view sv(s);

        int n = s.size();

        for(int i = 0; i < n - 10; i += 10){
            if(seen.count(sv.substr(i, 10)))
                res.insert(s.substr(i, 10));
            seen.insert(sv.substr(i, 10));
        }
    
        return {res.begin(), res.end()};
    }
};

File: 187. Repeated DNA Sequences.java

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        Set<String> res = new HashSet<>();
        Set<String> seen = new HashSet<>();
        int n = s.length();

        for(int i = 0; i <= n - 10; ++i){
            String subStr = s.substring(i, i + 10);
            if(seen.contains(subStr)){
                res.add(subStr);
            }
            seen.add(subStr);
        }

        return new ArrayList<>(res);
    }
}